### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by h33h**
Created Auth.storyboard
Created LoginScreen in Auth.storyboard
Created LoginScreenViewController and connected with storyboard
Created Presenter of LoginScreenViewController
Created custom textfield for AuthViewControllers
Setup LoginScreen in Auth.storyboard
Add realisation to LoginScreenViewController, Presenter of LoginScreenViewController, custom textfield
Some other edits

### CODE REVIEW

File: Mars Rover/Presenters/LoginFormPresenter.swift [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+class LoginFormPresenter {
+
+    // MARK: - LoginFormPresenter: Variables
+    private weak var delegate: LoginFormPresenterDelegate?
+}
+
+extension LoginFormPresenter {
+    // MARK: - LoginFormPresenter: Methods
```
Comment (IlyaValevich) [inline Mars Rover/Presenters/LoginFormPresenter.swift:24]: Follow for the uniformity of the code:
```
extension LoginFormPresenter {

     // MARK: - LoginFormPresenter: Methods
     // Delegate setter method
     func setDelegate(delegate: LoginFormPresenterDelegate) {
         self.delegate = delegate
     }

     // Login function will call some sign In methods from service
     func login() {
     }

}
```

File: Mars Rover/ViewControllers/LoginFormViewController.swift [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    }
+}
+
+extension LoginFormViewController: UITextFieldDelegate {
+    // MARK: - LoginFormViewController: UITextField Delegate Methods
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        // when click on return button, keyboard closes
+        textField.resignFirstResponder()
+        return true
+    }
```
Comment (IlyaValevich) [inline Mars Rover/ViewControllers/LoginFormViewController.swift:88]: Do not spam the code with unnecessary comments

File: Mars Rover/ViewControllers/LoginFormViewController.swift [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        else { return }
+        setElementsDisabled(value: false)
+        // present mainMenuVC
+        present(mainMenuVC, animated: true, completion: nil)
+    }
+    // if login failed
+    func failureLogin(error: String) {
+        setElementsDisabled(value: false)
+    }
+}
```
Comment (IlyaValevich) [inline Mars Rover/ViewControllers/LoginFormViewController.swift:80]:
```
  extension LoginFormViewController: LoginFormPresenterDelegate {

      // MARK: - LoginFormViewController: LoginFormPresenterDelegate Methods
      // if login succeed
      func successfullLogin() {
         // storyboard that contains MainMenuViewController
         let mainMenuStoryboard = UIStoryboard(name: "MainMenu", bundle: nil)
         // instantinate MainMenuViewController from mainMenuStoryboard
         guard let mainMenuVC = mainMenuStoryboard.instantiateViewController(withIdentifier: "MainMenu") as? MainMenuViewController
          else { return }
          setElementsDisabled(value: false)
          // present mainMenuVC
          present(mainMenuVC, animated: true, completion: nil)
      }

      // if login failed
      func failureLogin(error: String) {
          setElementsDisabled(value: false)
      }

  }
  ```

File: Mars Rover/ViewControllers/LoginFormViewController.swift (hunk 2 above)
Comment (IlyaValevich) [inline Mars Rover/ViewControllers/LoginFormViewController.swift:88]:
```
extension LoginFormViewController: UITextFieldDelegate {

     // MARK: - LoginFormViewController: UITextField Delegate Methods
     func textFieldShouldReturn(_ textField: UITextField) -> Bool {
         // when click on return button, keyboard closes
         textField.resignFirstResponder()
         return true
      }

}
```

File: Mars Rover/Views/AuthorizationTextField.swift [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    private var leftSideImageView = UIImageView()
+    // if leftSideImageView not have image isleftSideImageisHidden will be false
+    private var isleftSideImageisHidden: Bool = true {
+        didSet {
+            leftSideImageView.isHidden = isleftSideImageisHidden
+        }
+    }
+
+    // MARK: - AuthorizationTextField: Inspectable Variables
+    /// IBInspectable for realtime update of values in InterfaceBuilder
```
Comment (IlyaValevich) [inline Mars Rover/Views/AuthorizationTextField.swift:27]: ```// IBInspectable for realtime update of values in InterfaceBuilder```

File: Mars Rover/Views/AuthorizationTextField.swift [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+extension AuthorizationTextField {
+    // MARK: - AuthorizationTextField: Override Methods
+    override func textRect(forBounds bounds: CGRect) -> CGRect {
+        bounds.inset(by: insetsWithMargin)
+    }
+    override func editingRect(forBounds bounds: CGRect) -> CGRect {
+        bounds.inset(by: insetsWithMargin)
+    }
+}
```
Comment (IlyaValevich) [inline Mars Rover/Views/AuthorizationTextField.swift:92]:
```
extension AuthorizationTextField {

    // MARK: - AuthorizationTextField: Override Methods
    override func textRect(forBounds bounds: CGRect) -> CGRect {
        bounds.inset(by: insetsWithMargin)
    }

    override func editingRect(forBounds bounds: CGRect) -> CGRect {
        bounds.inset(by: insetsWithMargin)
    }

 }
```

File: Mars Rover/Views/AuthorizationTextField.swift [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        leftSideImageView.contentMode = .center
+        leftSideImageView.frame = CGRect(x: -8,
+                                         y: -8,
+                                         width: bounds.height + 16,
+                                         height: bounds.height + 16)
+        leftSideImageView.layer.cornerRadius = leftSideImageView.bounds.width / 2
+        leftSideImageView.isHidden = isleftSideImageisHidden
+        addSubview(leftSideImageView)
+    }
+}
```
Comment (IlyaValevich) [inline Mars Rover/Views/AuthorizationTextField.swift:82]:
```
extension AuthorizationTextField {

      // MARK: - AuthorizationTextField: Methods
      // called from init methods
      private func setupView() {
          // leftSideImageView setup
          let margin: CGFloat = 16
          leftSideImageView.contentMode = .center
          leftSideImageView.frame = CGRect(x: -margin / 2,
                                           y: -margin / 2,
                                           width: bounds.height + margin,
                                           height: bounds.height + margin)
          leftSideImageView.layer.cornerRadius = leftSideImageView.bounds.width / 2
          leftSideImageView.isHidden = isleftSideImageisHidden
          addSubview(leftSideImageView)
      }

 }
```

**Review by IlyaValevich — CHANGES_REQUESTED** [review]

File: Mars Rover/Presenters/LoginFormPresenter.swift [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    // Delegate setter method
+    func setDelegate(delegate: LoginFormPresenterDelegate) {
+        self.delegate = delegate
+    }
+    // Login function will call some sign In methods from service
+    func login() {
```
Comment (IlyaValevich) [inline Mars Rover/Presenters/LoginFormPresenter.swift:30]: Don't leave it empty, add // TODO:...

**Review by IlyaValevich — CHANGES_REQUESTED** [review]
The user now always gets to the authentication screen, shouldn't there be a switch to the menu for an already authenticated user?

**Comment by h33h** [comment]
> The user now always gets to the authentication screen, shouldn't there be a switch to the menu for an already authenticated user?

It should be realized as soon as authorization service will be added

File: Mars Rover/Presenters/LoginFormPresenter.swift [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+//
+
+import Foundation
+
+protocol LoginFormPresenterDelegate: AnyObject {
+
+    // MARK: - LoginFormPresenterDelegate: Delegate Methods
+    func successfullLogin()
+    func failureLogin(error: String)
+}
```
Comment (IlyaValevich) [inline Mars Rover/Presenters/LoginFormPresenter.swift:15]: Add empty line

File: Mars Rover/Presenters/LoginFormPresenter.swift [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+class LoginFormPresenter {
+
+    // MARK: - LoginFormPresenter: Variables
+    private weak var delegate: LoginFormPresenterDelegate?
+}
```
Comment (IlyaValevich) [inline Mars Rover/Presenters/LoginFormPresenter.swift:21]: Add empty line

File: Mars Rover/Views/AuthorizationTextField.swift [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    // MARK: - AuthorizationTextField: Methods
+    // called from init methods
+    private func setupView() {
+        // leftSideImageView setup
+        leftSideImageView.contentMode = .center
+        leftSideImageView.frame = CGRect(x: -8,
+                                         y: -8,
+                                         width: bounds.height + 16,
+                                         height: bounds.height + 16)
+        leftSideImageView.layer.cornerRadius = leftSideImageView.bounds.width / 2
```
Comment (IlyaValevich) [inline Mars Rover/Views/AuthorizationTextField.swift:82]: Use a variable instead of numbers

**Pull request merged**